reuseport.go

```diff
@@ -38,19 +38,28 @@ func reuseErrShouldRetry(err error) bool {
 	}
 }
 
-// Dials using reusport and then redials normally if that fails.
-func reuseDial(ctx context.Context, laddr *net.TCPAddr, network, raddr string) (net.Conn, error) {
+// Dials using reuseport and then redials normally if that fails.
+// TODO(anacrolix): This shouldn't fail anymore: Remove fallback.
+func reuseDial(ctx context.Context, laddr *net.TCPAddr, network, raddr string) (con net.Conn, err error) {
 	if laddr == nil {
 		return fallbackDialer.DialContext(ctx, network, raddr)
 	}
 
-	d := reuseport.Dialer{
-		D: net.Dialer{
-			LocalAddr: laddr,
-		},
+	d := net.Dialer{
+		LocalAddr: laddr,
+		Control:   reuseport.Control,
 	}
+	defer func() {
+		if err != nil {
+			return
+		}
+		// This is transplanted from go-reuseport, which once set no linger on
+		// dialing and may be a requirement for desired behaviour in this
+		// package.
+		con.(*net.TCPConn).SetLinger(0)
+	}()
```

Comment:
IIRC, if we don't do this we get weird issues where we can't disconnect, restart, and then reconnect (never was sure why).

Reply:
I believe it's because even though SO_REUSEPORT or equivalent, when dialing you can't reuse the same address tuple. Lingering means an old socket that still has pending writes will still occupy the address tuple. Linger(0) means that as soon as the last file descriptor referencing it is closed, the socket and its address tuple is removed from the network table without waiting for pending writes.